File: IAIso-v5.0/core/iaiso-csharp/src/Iaiso.Middleware/Bedrock.cs
using System.Collections.Generic;
using System.Text.Json.Nodes;
using Iaiso.Core;

namespace Iaiso.Middleware.Bedrock;

/// <summary>
/// IAIso wrapper for AWS Bedrock runtime. Supports both Converse API
/// (preferred — normalized usage extraction) and the lower-level
/// InvokeModel API.
/// </summary>
public interface IBedrockClient
{
    ConverseResponse Converse(JsonObject @params);
    InvokeResponse InvokeModel(JsonObject @params);
}

public sealed class ConverseUsage
{
    public long InputTokens { get; set; }
    public long OutputTokens { get; set; }
    public long TotalTokens { get; set; }
}

public sealed class ConverseContentBlock
{
    public bool HasToolUse { get; }
    public ConverseContentBlock(bool hasToolUse) { HasToolUse = hasToolUse; }
}

public sealed class ConverseResponse
{
    public ConverseUsage Usage { get; }
    public IReadOnlyList<ConverseContentBlock> Content { get; }
    public ConverseResponse(ConverseUsage? usage, IReadOnlyList<ConverseContentBlock>? content)
    {
        Usage = usage ?? new ConverseUsage();
        Content = content ?? new List<ConverseContentBlock>();
    }
}

public sealed class InvokeResponse
{
    public string ModelId { get; }
    public byte[] Body { get; }
    public InvokeResponse(string modelId, byte[]? body)
    {
        ModelId = modelId;
        Body = body ?? System.Array.Empty<byte>();
    }
}

public sealed class BedrockOptions
{
    public bool RaiseOnEscalation { get; set; }
    public static BedrockOptions Defaults() => new();
}

public sealed class BoundedBedrockClient
{
    private readonly IBedrockClient _raw;
    private readonly BoundedExecution _execution;
    private readonly BedrockOptions _opts;

    public BoundedBedrockClient(IBedrockClient raw, BoundedExecution execution, BedrockOptions? opts = null)
    {
        _raw = raw;
        _execution = execution;
        _opts = opts ?? BedrockOptions.Defaults();
    }

    public ConverseResponse Converse(JsonObject @params)
    {
        CheckState();
        ConverseResponse resp;
        try { resp = _raw.Converse(@params); }
        catch (System.Exception e) { throw new MiddlewareException.Provider(e.Message, e); }

        long tokens = resp.Usage.TotalTokens;
        if (tokens == 0) tokens = resp.Usage.InputTokens + resp.Usage.OutputTokens;
        long toolCalls = 0;
        foreach (var b in resp.Content) if (b.HasToolUse) toolCalls++;
        string modelId = @params["modelId"]?.GetValue<string>() ?? "unknown";
        _execution.RecordStep(new StepInput
        {
            Tokens = tokens,
            ToolCalls = toolCalls,
            Tag = $"bedrock.converse:{modelId}",
        });
        return resp;
    }

    public InvokeResponse InvokeModel(JsonObject @params)
    {
        CheckState();
        InvokeResponse resp;
        try { resp = _raw.InvokeModel(@params); }
        catch (System.Exception e) { throw new MiddlewareException.Provider(e.Message, e); }

        string modelId = !string.IsNullOrEmpty(resp.ModelId)
            ? resp.ModelId
            : (@params["modelId"]?.GetValue<string>() ?? "unknown");
        // Best-effort: model-specific bodies require user adapter to extract token counts.
        _execution.RecordStep(new StepInput
        {
            Tag = $"bedrock.invokeModel:{modelId}",
        });
        return resp;
    }

    private void CheckState()
    {
        var pre = _execution.Check();
        if (pre == StepOutcome.Locked) throw new MiddlewareException.Locked();
        if (pre == StepOutcome.Escalated && _opts.RaiseOnEscalation)
            throw new MiddlewareException.EscalationRaised();
    }
}

File: IAIso-v5.0/core/iaiso-ruby/lib/iaiso/middleware/bedrock.rb
# frozen_string_literal: true

require_relative "errors"
require_relative "../core/pressure_engine"
require_relative "../core/step_outcome"

module IAIso
  module Middleware
    # AWS Bedrock middleware. Supports Converse (preferred) and the
    # lower-level InvokeModel API.
    module Bedrock
      ConverseUsage = Data.define(:input_tokens, :output_tokens, :total_tokens) do
        def self.empty = new(input_tokens: 0, output_tokens: 0, total_tokens: 0)
      end
      ConverseContentBlock = Data.define(:has_tool_use)
      ConverseResponse = Data.define(:usage, :content)
      InvokeResponse = Data.define(:model_id, :body) do
        def self.empty = new(model_id: "", body: "")
      end

      Options = Data.define(:raise_on_escalation) do
        def self.defaults = new(raise_on_escalation: false)
      end

      class BoundedClient
        def initialize(raw:, execution:, options: Options.defaults)
          @raw = raw
          @execution = execution
          @options = options
        end

        def converse(params)
          check_state!
          resp =
            begin
              @raw.converse(params)
            rescue StandardError => e
              raise ProviderError, e.message
            end
          tokens = resp.usage.total_tokens.to_i
          tokens = resp.usage.input_tokens.to_i + resp.usage.output_tokens.to_i if tokens.zero?
          tool_calls = resp.content.count { |b| b.has_tool_use }
          model_id = params[:modelId] || params["modelId"] || "unknown"
          @execution.record_step(IAIso::Core::StepInput.build(
            tokens: tokens, tool_calls: tool_calls,
            tag: "bedrock.converse:#{model_id}",
          ))
          resp
        end

        def invoke_model(params)
          check_state!
          resp =
            begin
              @raw.invoke_model(params)
            rescue StandardError => e
              raise ProviderError, e.message
            end
          model_id = !resp.model_id.to_s.empty? ? resp.model_id : (params[:modelId] || params["modelId"] || "unknown")
          @execution.record_step(IAIso::Core::StepInput.build(
            tag: "bedrock.invokeModel:#{model_id}",
          ))
          resp
        end

        private

        def check_state!
          pre = @execution.check
          raise LockedError if pre == IAIso::Core::StepOutcome::LOCKED
          if pre == IAIso::Core::StepOutcome::ESCALATED && @options.raise_on_escalation
            raise EscalationRaisedError
          end
        end
      end
    end
  end
end
